jetbrains-gateway/it/software/aws/toolkits/jetbrains/gateway/DevEnvConnectTest.kt

```diff
@@ -0,0 +1,263 @@
+                    CawsConnectionParameters.CAWS_PROJECT to environment.projectName,
+                    CawsConnectionParameters.CAWS_ENV_ID to environment.id,
+                ),
+                ConnectionRequestor.Local
+            )
+        } ?: error("null connection handle")
+
+        yieldAll(
+            listOf(
+                test(::`wait for environment ready`),
+                test(::`poll for bootstrap script availability`),
+                test(::`wait for backend start`),
+                test(::`wait for backend connect`)
+            )
+        )
+    }.iterator()
+
+    fun `wait for environment ready`() = runBlocking {
+        waitUntil(
+            succeedOn = {
+                it.status() == DevEnvironmentStatus.RUNNING
+            },
+            failOn = {
+                it.status() == DevEnvironmentStatus.FAILED || it.status() == DevEnvironmentStatus.DELETED
+            },
+            maxDuration = Duration.ofMinutes(2),
+            call = {
+                client.getDevEnvironment {
+                    it.spaceName(environment.spaceName)
+                    it.projectName(environment.projectName)
+                    it.id(environment.id)
+                }
+            }
+        )
+    }
+
+    fun `poll for bootstrap script availability`() = runBlocking {
+        waitUntil(
+            succeedOn = {
+                it == StdOutResult.SUCCESS
+            },
+            failOn = {
+                connectionHandle.lifetime.isNotAlive
+            },
+            maxDuration = Duration.ofMinutes(5),
+            call = {
+                // condition looks inverted because we want failure if script not found
+                ssmFactory
+                    .executeCommandNonInteractive("sh", "-c", "test -z \"\$(find /tmp -name \"start-ide.sh\" 2>/dev/null)\" && echo false || echo true")
+                    .resultFromStdOut()
+            }
+        )
+    }
+
+    fun `wait for backend start`() = runBlocking {
+        // make sure port forward is alive
+        lazyPortForward.value
+
+        waitUntil(
+            succeedOn = {
+                it != null
+            },
+            failOn = { connectionHandle.lifetime.isNotAlive },
+            maxDuration = Duration.ofMinutes(5),
+            call = {
+                tryOrNull {
+                    HttpRequests.request(endpoint).readString()
+                }
+            }
+        )
+    }
+
+    fun `wait for backend connect`() = runBlocking {
+        waitUntil(
+            succeedOn = { status ->
+                status.projects?.any { it.users.size > 1 } == true
+            },
+            failOn = { connectionHandle.lifetime.isNotAlive },
+            maxDuration = Duration.ofMinutes(5),
+            call = {
+                UnattendedHostStatus.fromJson(HttpRequests.request(endpoint).readString())
+            }
+        )
+
+        ThinClientTrackerService.getInstance().closeThinClient(environment.id)
```

Comment:
Should this be in after?

Reply:
dont think it matters. the client handle itself isn't available until it has successfully connected